src/main/scala/com/cognite/spark/datasource/CdpRelation.scala

```diff
@@ -22,31 +33,66 @@ case class DeleteItem(
     id: Long
 )
 
-abstract class CdpRelation[T: Decoder](config: RelationConfig, shortName: String)
+case class PushdownFilter(fieldName: String, value: String)
+
+abstract class CdpRelation[T <: Product: Decoder](config: RelationConfig, shortName: String)
     extends BaseRelation
     with TableScan
     with Serializable
+    with PrunedFilteredScan
     with CdpConnector {
   @transient lazy private val itemsRead =
     MetricsSource.getOrCreateCounter(config.metricsPrefix, s"$shortName.read")
 
+  def fieldsWithPushdownFilter: Seq[String] = Seq[String]()
+
   val sqlContext: SQLContext
-  override def buildScan(): RDD[Row] =
+  override def buildScan(): RDD[Row] = buildScan(Array.empty, Array.empty)
+
+  override def buildScan(requiredColumns: Array[String], filters: Array[Filter]): RDD[Row] = {
+    val pushdownFilters: Seq[PushdownFilter] = fieldsWithPushdownFilter
+      .map(f => (f, filters.flatMap(getFilter(_, f)))) // map from fieldname to its filtering values
+      .flatMap(pdf => pdf._2.map(v => PushdownFilter(pdf._1, v)))
+      .distinct
+
     CdpRdd[T](
       sqlContext.sparkContext,
       (e: T) => {
         if (config.collectMetrics) {
           itemsRead.inc()
         }
-        toRow(e)
+        if (requiredColumns.isEmpty) toRow(e) else toRow(e, requiredColumns)
       },
       listUrl(),
       config,
+      pushdownFilters,
       cursors()
     )
+  }
 
   def toRow(t: T): Row
 
+  def toRow[A <: Product](item: A, requiredColumns: Array[String]): Row = {
+    val values = item.productIterator
+    val itemMap = item.getClass.getDeclaredFields.map(_.getName -> values.next).toMap
+
+    Row.fromSeq(requiredColumns.map(itemMap(_)).toSeq)
```

Comment:
This function doesn't work for higher kinded types (Option, Scala Seq etc) I believe?

Reply:
Which means that requiredColumns has to point to primitives in a case class / class.

Reply:
Seems to work fine, requiredColumns is usually the name of all columns and all tests are still passing.

Reply:
Hmm, last time I checked it didn't create a row as I expected. However if that's the case, we can remove the macro code for creating a row, though the macro code will probably be faster since it unrolls a dynamic loop at compile time.

Reply:
And given that toRow will potentially be called for billions of rows, that type of performance improvement actually matters. However, performance measuring is something we should add to our todo list.

Reply:
I agree. `SparkSchemaHelperImpl.asRow` is still called when scanning without pushdown filters, so let's keep it.